refactor: apply a bunch of lints

EdJoPaTo opened the pull request:
In the beginning I wanted to add `const` improvements to widgets I use, but then I thought I should apply a bunch of lints first before doing more refacting like `const` new methods.
Applying them to the whole project is similar in comparison to just a few files, so I did that.

I ran my personal bunch of lints over the entire project (pedantic, nursery and some restricted) and applied some of them. I went through the lints and checked what is at least partially done already in this repo to ensure a consistent state. For example [`use_self`](https://rust-lang.github.io/rust-clippy/master/index.html#/use_self) is partially used in this repo. The lint ensures its done everywhere consistently. The same with the other lints.
Currently, lints which did not trigger at least some kind of refactoring are not added. In the long run I would like to enable `clippy::pedantic` (except lints like [`module_name_repetitions`](https://rust-lang.github.io/rust-clippy/master/index.html#/module_name_repetitions)) for this project.

This creates a lot of changes but except the change to the `WidgetRef` it should not be a breaking change. (It somewhat amazed me that the `WidgetRef` API required ownership and not a reference or have I overlooked something? It's in its own commit)

This probably creates some churn with other PR created before this is merged.

Looking at all the changes together is somewhat pointless. Each commit should be checked on its own. It should also be considered if this should be squashed on merge like the other PR do normally. I think it's useful to keep the commits individually. But that is not my decision.

I can rebase this and add/improve/remove commits/lints. I can also cherry-pick some into their own PR to fast-lane them. We should discuss the lints.

joshka reviewed (commented):
This looks good to me. I agree with all the changes generally (with a couple of questions / comments). I ended up just looking at the entire PR as I saw your comment about looking at each commit after I reviewed.

I like your thoughts on merging this as multiple commits generally, but I think I'm also fine with squashing the changes as this fits better with the workflow (and avoids having to split this into multiple PRs). There's only a few outstanding PRs right now, and the benefit of this PR outweighs the extra work to rebase them if needed.

I recall that there's a way to mark specific commits not to be included when annotating the file with git blame - perhaps we need that for this PR?

P.s. thanks for all the effort on this one - much appreciated!

EdJoPaTo commented on CONTRIBUTING.md:
```diff
@@ -112,7 +112,7 @@ exist to show coverage directly in your editor. E.g.:
 ### Documentation
 
-Here are some guidelines for writing documentation in Ratatui.  
+Here are some guidelines for writing documentation in Ratatui.
```
I think we should prefer explicit new lines then rather than some quirky Markdown syntax. Markdown implementations differ a lot which also does not help with these smaller features.

New lines would end up creating a new paragraph so it should be either without new line or as a new paragraph in my opinion.

joshka replied:
I agree - here this line needs to be a new paragraph, but the lines below should be re-wrapped at 100 characters so the next line end up being at the end of the line where the double space was.

joshka replied:
Added suggestion https://github.com/ratatui-org/ratatui/pull/974/files/53f29f5b5fa74a10d47eebf36f66f2048bde7e71#r1509934442

EdJoPaTo replied:
I applied the new wrapping in the way I think it was intended. Can you check if its correct?

EdJoPaTo commented on examples/scrollbar.rs:
```diff
@@ -107,6 +110,7 @@ fn run_app<B: Backend>(
     }
 }
 
+#[allow(clippy::too_many_lines, clippy::cast_possible_truncation)]
```
In the long run, yes.
But then we can search for allow statements. There are more lints like cast related or potentially wrong or panicking arithmetic which should be enabled and fixed but that’s way easier to on a smaller scope with more refactorings.

For this PR I’m concerned about enabling more general lints and doing only minor refactoring unrelated to the direct lints.

joshka replied:
Sounds good

EdJoPaTo commented on examples/vhs/custom_widget.tape:
```diff
@@ -18,4 +18,4 @@ Space
 Left
-Space
\ No newline at end of file
+Space
```
There was no blank line in the end and now there is. I ensured that in the editorconfig and it’s now the same in all the files.

It creates weird diffs which is why I prefer to have them always.

joshka replied:
I prefer having them always too. I think I misread this diff as removing the end of file newline instead of adding them.

EdJoPaTo commented on src/lib.rs:
```diff
@@ -1,4 +1,22 @@
 #![forbid(unsafe_code)]
+#![warn(
```
I like that you are bringing this up! I briefly thought about that.

It’s probably not even needed to bump the MSRV as they should be ignored on older versions where they are still unknown? Lints will be checked on stable and for older versions unknown-lints should be ignored anyway.

Also it would affect examples then too I think?

joshka replied:
Our MSRV policy is 2 versions behind stable - mainly because that matches policies of our deps - so 1.74.0 is fine here. Can you take a look at whether this can be set without a tonne of work on the examples - if it can, then let's do it there.

joshka replied:
I'm going to mark this one as resolved - we can do this as a secondary step if needed.

EdJoPaTo replied:
I waited with this until #976 is merged :innocent:

EdJoPaTo replied:
Done :sparkles:

EdJoPaTo commented on src/widgets/block/padding.rs:
```diff
@@ -186,7 +186,7 @@ mod tests {
 
     #[test]
-    fn can_be_const() {
+    const fn can_be_const() {
```
Yeah this case is funny. There are other can_be_const test fn which can not be const which is even more funny (because of asserts).
It’s basically pointless to have const in tests as it doesn’t bring runtime performance but it satisfies the lint. For tests I would not care about removing const when not needed.

EdJoPaTo commented on src/widgets/scrollbar.rs:
```diff
@@ -1,3 +1,12 @@
+#![warn(clippy::pedantic)]
+#![allow(
+    clippy::cast_possible_truncation,
+    clippy::cast_precision_loss,
+    clippy::cast_sign_loss,
```
See https://github.com/ratatui-org/ratatui/pull/974#discussion_r1505517664

EdJoPaTo commented on src/widgets/table/table.rs:
```diff
@@ -223,16 +223,16 @@ pub struct Table<'a> {
-            highlight_symbol: Default::default(),
-            highlight_spacing: Default::default(),
+            block: None,
+            style: Style::new(),
+            highlight_style: Style::new(),
```
Default can not be const so I personally would prefer using const new when possible. That should allow for more compiler optimisations.

(no it shouldn’t trigger a lint)

There are more cases where new over default would allow for more const optimisations but I don’t think that’s part of PR as it includes more refactoring.

joshka replied:
This is not a blocking comment - so can be left alone or fixed at your preference. marking it as resolved for now.

EdJoPaTo commented on src/terminal/frame.rs:
````diff
@@ -89,10 +89,10 @@ impl Frame<'_> {
     /// ```
     #[stability::unstable(feature = "widget-ref")]
-    pub fn render_widget_ref<W: WidgetRef>(&mut self, widget: W, area: Rect) {
+    pub fn render_widget_ref<W: WidgetRef>(&mut self, widget: &W, area: Rect) {
````
W can also contain `&WidgetRef` or `&mut WidgetRef` :thinking: 

https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md/#1-t-only-contains-owned-types

EdJoPaTo replied:
```rust
impl Widget for &Something {…}
```

This means that a reference of something is enough to render… no need to transfer the ownership… I think this is a bigger topic that should be moved out of this PR?

joshka replied:
The main reason this is unstable is because there's some possible options for implementing this (noted at https://users.rust-lang.org/t/calling-a-method-with-a-self-parameter-on-box-dyn-trait/106220) that would be slightly incompatible.

The history on this is that Widget takes `self` instead of &self, and changing that breaks everything, so we needed something more interesting to make &self and Box<dyn Widget> to work. It's possible that we didn't actually need something more interesting in the long run. It's also possible that we could just break everything instead.

The relevant PRs for more context are:
- https://github.com/ratatui-org/ratatui/pull/833
- https://github.com/ratatui-org/ratatui/pull/903
- https://github.com/ratatui-org/ratatui/pull/897

EdJoPaTo replied:
I allowed this lint on this. It should be discussed in its own place. When searching for allowed lints this will come up again.

joshka commented on CONTRIBUTING.md:
```diff
@@ -112,7 +112,7 @@ exist to show coverage directly in your editor. E.g.:
 ### Documentation
 
-Here are some guidelines for writing documentation in Ratatui.  
+Here are some guidelines for writing documentation in Ratatui.
```
```suggestion
Here are some guidelines for writing documentation in Ratatui.

```

joshka commented on src/style/color.rs:
```diff
@@ -420,11 +420,11 @@ mod tests {
-        assert_eq!(Color::from_u32(0x00FF00), Color::Rgb(0, 255, 0));
-        assert_eq!(Color::from_u32(0x0000FF), Color::Rgb(0, 0, 255));
-        assert_eq!(Color::from_u32(0xFFFFFF), Color::Rgb(255, 255, 255));
+        assert_eq!(Color::from_u32(0x0000_0000), Color::Rgb(0, 0, 0));
+        assert_eq!(Color::from_u32(0x00FF_0000), Color::Rgb(255, 0, 0));
+        assert_eq!(Color::from_u32(0x0000_FF00), Color::Rgb(0, 255, 0));
+        assert_eq!(Color::from_u32(0x0000_00FF), Color::Rgb(0, 0, 255));
+        assert_eq!(Color::from_u32(0x00FF_FFFF), Color::Rgb(255, 255, 255));
```
This change looks a bit weird (as these values are hex representations of colors, and not strictly pure u32 values). Do you think it's worth adding an allow for this?

joshka commented on src/style/palette/material.rs:
```diff
@@ -523,84 +523,300 @@ pub const GRAY: NonAccentedPalette = NonAccentedPalette::from_variants(variants:
 // Black and white included for completeness
-pub const BLACK: Color = Color::from_u32(0x000000);
-pub const WHITE: Color = Color::from_u32(0xFFFFFF);
+pub const BLACK: Color = Color::from_u32(0x0000_0000);
+pub const WHITE: Color = Color::from_u32(0x00FF_FFFF);
```
And here - same question - I think we should allow this format for the entire file as the 3 byte format is better than the 4 byte formats (search-ability / readability / vertical space / etc.).

joshka commented on src/style/palette/tailwind.rs:
```diff
@@ -293,360 +293,360 @@ pub struct Palette {
 #[rustfmt::skip]
 /// <style>.palette div{width:22rem;height:2rem}</style><div class="palette" style="display:flex;flex-direction:row"><div style="background-color: #000000"></div></div>
-pub const BLACK: Color = Color::from_u32(0x000000);
+pub const BLACK: Color = Color::from_u32(0x0000_0000);
```
Also here as for color / material.

joshka commented on src/widgets/canvas.rs:
```diff
@@ -71,10 +71,6 @@ struct Layer {
 /// Braille patterns will have a resolution of 2x4 dots per cell. This means that a grid of 10x10
 /// cells will have a resolution of 20x40 dots.
 trait Grid: Debug {
-    /// Get the width of the grid in number of terminal columns
-    fn width(&self) -> u16;
-    /// Get the height of the grid in number of terminal rows
-    fn height(&self) -> u16;
```
(Comment for rationale - no change needed) Initially this one looks a bit odd, but the trait is private, and so I'm guessing that these values are just not accessed, and so are not necessary.

joshka reviewed (changes requested):
LGTM (I re-reviewed all the changed files since last review line by line).

I'd like to undo the changes to the u32 literals in the Color files, because the 3 byte values are actually more readable than the 4 byte values Perhaps keep the warn in the lint config, but add `allow`s on the 3 files in question. https://github.com/ratatui-org/ratatui/pull/974/commits/d76a7894a0cfe4504af13d4ffad635e2f291162f is the commit in question. Does that make sense?

I plan to merge this as a rebase - do you think it would be worth adding the PR number to all the commit messages to make it easy to link back to this PR and the comments (for anyone spelunking the codebase at some later date)?

Perhaps as a last thing to add for this, would you mind creating a .git-blame-ignore-revs file that marks these commits from blame (to make it easier to understand the blame output)? See https://git-scm.com/docs/git-blame / https://www.michaelheap.com/git-ignore-rev/ (or other blog posts for details).

joshka commented:
P.s. this was really done well. You might consider writing an article / blog post on the approach you took to doing this in a controlled manner. Some things I'd really like to read about (and I'm sure others might feel the same) were:
- how - i.e. were you using an automated tool, iterating on cargo clippy --fix, using regex magic
- what / why? How were you picking the lints (just going down the pedantic list?) Which lints did you ignore based on experience and why?
- What problems did you run into while doing this?
- How long did it take
- How should the project move on from this - i.e. what's now enabled for all users.
- How easy was the process generally?
- What could make it better?
etc.

EdJoPaTo commented:
> Perhaps as a last thing to add for this, would you mind creating a .git-blame-ignore-revs file that marks these commits from blame (to make it easier to understand the blame output)? See https://git-scm.com/docs/git-blame / https://www.michaelheap.com/git-ignore-rev/ (or other blog posts for details).

Not sure how useful this is in practice (but I have never used it). Personally I like the browser blame view. It has a "view before this commit happened" button per line. That way I see the last change was commit X, it looks irrelevant, what happened before that? This approach worked well for me in the past.

Another thing is the rebase approach we are using, both when accepting this PR and when I update it. The commit hashes are changing often currently. When this is still wanted it should be added after accepting this PR?

EdJoPaTo commented:
@joshka I added more lints in more commits, the rest should be the same as before.
Starts with adding `clippy::should_panic_without_expect`, currently this range: [83e98b1..1c2e86a](https://github.com/ratatui-org/ratatui/compare/83e98b13d6a2f078b836f9102b908e90de0f822f..1c2e86a76a0080d6732baacfdc89158a14d6285b)

joshka commented:
> When this is still wanted it should be added after accepting this PR?

Yep - Makes sense as another PR once merged.

joshka commented:
I rebased the fixup commit and merged this.

EdJoPaTo commented:
> P.s. this was really done well. I wonder if you might might consider writing an article / blog post on the approach you took to doing this in a controlled manner. Some things I'd really like to read about (and I'm sure others might feel the same) were:
> 
>     * how - i.e. were you using an automated tool, iterating on cargo clippy --fix, using regex magic
> 
>     * what / why? How were you picking the lints (just going down the pedantic list?) Which lints did you ignore based on experience and why?
> 
>     * What problems did you run into while doing this?
> 
>     * How long did it take
> 
>     * How should the project move on from this - i.e. what's now enabled for all users.
> 
>     * How easy was the process generally?
> 
>     * What could make it better?
>       etc.

Guess the core things are… One by one and commit often.
Run all the lints, for example all pedantic and pick exactly one lint. Then only enable this lint and fix it. Then commit. If something went wrong (regex replaces for example are easy to mess up) or is more tedious than expected then revert the changes back to the last commit.

Personally I started with clippy in general, then pedantic, then nursery and then seeking additional restricted lints which fit my style well. I also like to adapt my style in order to work with a lint which is helping me instead of fighting it (wildcard imports in this repo for example).

Clippy lints are really well done with their hints (and sometimes automatic fixes). They supply quick feedback on what to improve and why. Fast feedback results in quick learning how to write better Rust.

How should users continue… a good idea is to enable clippy for your editor. For example with VS Code set the Check Command to `clippy`. Then you get quick feedback while writing code. In the beginning its annoying (remember, the lints are called pedantic for a reason) but later on I tend to write better code directly. Or know that the lint will come up and use its suggestion afterwards. Like `.map(|o| o.width())` is easy to write quickly, the lint will then suggest to use `.map(Line::width)` and just apply that. It's similar to `cargo fmt`. I know how to format it, but it's easier to write it the easy way and let the command enforce the "correct" way.

In the future the allowed lints in the `Cargo.toml` should be removed and fixed or allowed on the individual cases. Allowed lints in the code should be checked upon.

How easy was it? Clippy output is really helpful!

How long did it take? Definitely many hours as some lints were attempted, started to be refactored and then aborted due to some reason in some places. Like `must_use_candidate` or `same-name-method`. `same-name-method` is annoyed about style and Stylize methods which resulted looking into creating a derive macro for Stylize in order to have `const` `.red()` methods which side tracked me and consumed even more time. Not all lints should be directly applied, some need more refactoring to do it right.

Which lints have I picked? Ran my personal list, picked one that seems suitable based on experience and applied it. `use_self` for example was already partially done in this repo and can be `--fix`ed so its an easy start. (Hopefully it would have detected and fixed all cases in this repo in the future automatically: https://github.com/rust-lang/rust-clippy/issues/12381). Personally I wanted to have `const fn` so `missing_const_for_fn` was also high on the list. At some point it went into "get all the pedantic ones" and less "which seems like a low hanging fruit".

Automated tools? Not really? Depends heavily on the lint. Sometimes `--fix` helps, sometimes regex search replace over all the files, sometimes multi cursor / macros. But as always, the more you edit, the more "oh well" can happen. (→ commit often)

What would make it better? Probably the interest in clean and idiomatic code in more people? More lints should be enabled by default especially for libraries. Especially stuff like const moves from libraries into the end user code.

Another consideration: `clippy.toml`: `avoid-breaking-exported-api = false`. Breaking changes on public API can result in performance improvements and more idiomatic, less confusing code down the line.
